[Merged by Bors] - chore(Archive): golf and generalise ascending-descending sequences

b-mehta opened the pull request:
This proof was 5 years old, not in a good generality, and overly long. We fix the latter two points.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta committed:
wip

b-mehta committed:
Merge branch 'master' into golf-erdos-szekeres

b-mehta committed:
tidy

b-mehta committed:
move lemmas

b-mehta committed:
fixup

grunweg reviewed (commented):
Thanks for cleaning up this old corner of mathlib! I have a few superficial comments about further golfing/clean-up opportunities --- some of these are subjective, perhaps others are useful. (I just realise this is still a draft: I hope you don't mind the comments anyway. Sorry if you did!)

b-mehta commented on Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean:
```diff
@@ -23,12 +25,92 @@ https://en.wikipedia.org/wiki/Erdos-Szekeres_theorem#Pigeonhole_principle.
+    · refine ⟨?_, ht₁⟩
+      intro x hx hxj
+      exact (ht₁.monotoneOn hx hti.1 (hti.2 hx)).trans_lt hfij
+    · intro x hx
+      exact (this x hx).le
```
```suggestion
    · exact fun x hx ↦ (this x hx).le
```

b-mehta committed:
Apply suggestions from code review

b-mehta commented on Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean:
```diff
@@ -23,12 +25,92 @@ https://en.wikipedia.org/wiki/Erdos-Szekeres_theorem#Pigeonhole_principle.
+  obtain ⟨t, hti, ht₁, ht₂⟩ := this
+  simp only [incSequencesTo, mem_image, mem_filter, mem_univ, true_and, exists_and_right,
+    and_assoc]
+  have : ∀ x ∈ t, x < j := by
```
Something like this should work, but I prefer it as is for the readability of "everything in t is below j"

b-mehta commented on Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean:
```diff
@@ -23,12 +25,92 @@ https://en.wikipedia.org/wiki/Erdos-Szekeres_theorem#Pigeonhole_principle.
+    · intro x hx
+      exact (this x hx).le
+  have : j ∉ t := by
+    intro hj
```
done

b-mehta commented on Archive/Wiedijk100Theorems/AscendingDescendingSequences.lean:
```diff
@@ -23,12 +25,92 @@ https://en.wikipedia.org/wiki/Erdos-Szekeres_theorem#Pigeonhole_principle.
+    · refine ⟨?_, ht₁⟩
+      intro x hx hxj
+      exact (ht₁.monotoneOn hx hti.1 (hti.2 hx)).trans_lt hfij
+    · intro x hx
```
done

b-mehta committed:
Merge branch 'master' into golf-erdos-szekeres

b-mehta committed:
Merge branch 'master' into golf-erdos-szekeres

b-mehta committed:
Merge branch 'master' into golf-erdos-szekeres

b-mehta committed:
fix lint

j-loreaux reviewed (commented):
The only downside here is that some of the comments were lost from the original proof. I realize that the new approach makes things significantly less complicated, but maybe you could add a few sentences as comments?

bors d+

b-mehta committed:
Merge branch 'master' into golf-erdos-szekeres

b-mehta committed:
add comments

b-mehta commented:
> The only downside here is that some of the comments were lost from the original proof. I realize that the new approach makes things significantly less complicated, but maybe you could add a few sentences as comments?

I've added some more docstrings, and added some signposting sentences in the proof!

bors merge
